File: Exercises2024/Chapter06_vle_free_energy/Fortran/1-Widom/Source/readdat.f
      Subroutine Readdat(NumberOfEquilibrationCycles,NumberOfProductionCycles,
     &     SamplingFrequency,NumberOfDisplacementsPerCycle,MaxDisplacement)

      Implicit None

C     ---Read Input Data And Model Parameters
 
      Include 'parameter.inc'
      Include 'system.inc'
      Include 'potential.inc'
      Include 'conf.inc'

      Integer InitType,NumberOfEquilibrationCycles,NumberOfProductionCycles,
     &     I,NumberOfDisplacementsPerCycle,SamplingFrequency,Sstmm
      Double Precision Eps, Sig, Boxf, Rhof, Rho, MaxDisplacement, M1
 
C     ---Read Simulation Data
C
C     InitType = 0 : Initialize From A Lattice
C                1 : Read Configuration From Disk
C     Rho      = Density

      Open (15,FILE='input',FORM='FORMATTED')

      Read (15, *)
      Read (15, *) InitType,NumberOfEquilibrationCycles,
     &     NumberOfProductionCycles,SamplingFrequency
      Read (15, *)
      Read (15, *) MaxDisplacement
      Read (15, *)
      Read (15, *) NumberOfDisplacementsPerCycle
      Read (15, *)
      Read (15, *) NumberOfParticles,Temperature,Rho
      Read (15, *)
      Read (15, *) Eps, Sig, Mass, Rc

C     ---Initialize Random Number Generator
      
      M1 = 0.001d0*Dble(Mod((10+10*Sstmm()),1000))

      If(M1.Lt.0.001d0) M1 = 0.001d0
      If(M1.Gt.0.999d0) M1 = 0.999d0

      Call InitializeRandomNumberGenerator(M1)
 
      If (NumberOfParticles.Gt.MaxParticles) Then
         Write (6, *) ' Error: Number Of Particles Too Large'
         Stop
      End If

      Box  = (Dble(NumberOfParticles)/Rho)**(1.D0/3.D0)
      HalfBox = Box/2

C     ---Read/Generate Configuration

      If (InitType.Eq.0) Then

C        ---Generate Configuration Form Lattice

         Call Lattice
      Else
         Write (6, *) ' Read Conf From Disk '
         Open (11,FILE='restart.dat',FORM='FORMATTED')
         Read (11, *) Boxf
         Read (11, *) NumberOfParticles
         Read (11, *) MaxDisplacement
         Rhof = Dble(NumberOfParticles)/Boxf**3
         If (Abs(Boxf-Box).Gt.1d-6) Then
            Write (6, 99007) Rho, Rhof
         End If
         Do I = 1, NumberOfParticles
            Read (11, *) X(I), Y(I), Z(I)
            X(I) = X(I)*Box/Boxf
            Y(I) = Y(I)*Box/Boxf
            Z(I) = Z(I)*Box/Boxf
         End Do
         Close(11)
      End If

C     ---Write Input Data

      Write (6, 99001) NumberOfEquilibrationCycles, NumberOfProductionCycles, SamplingFrequency
      Write (6, 99002) NumberOfDisplacementsPerCycle, MaxDisplacement
      Write (6, 99003) NumberOfParticles, Temperature, Rho, Box
      Write (6, 99004) Eps, Sig, Mass

C     ---Calculate Parameters:

      Beta = 1/Temperature
      
C     ---Calculate Cut-Off Radius Potential

      Rc    = Min(Rc, HalfBox)
      Rc2   = Rc*Rc
      Eps4  = 4.0d0*Eps
      Eps48 = 48.0d0*Eps
      Sig2  = Sig*Sig
      
      Return

99001 Format ('  Number Of Equilibration Cycles             :', I10, /, 
     &        '  Number Of Production Cycles                :', I10, /, 
     &        '  Sample Frequency                           :', I10, /)
99002 Format ('  Number Of Att. To Displ. A Part. Per Cycle :', I10, /, 
     &        '  Maximum Displacement                       :', F10.3, 
     &        //)
99003 Format ('  Number Of Particles                        :', I10, /, 
     &        '  Temperature                                :', F10.3, 
     &        /, '  Density                                    :', 
     &        F10.3, /, '  Box Length                                 :'
     &        , F10.3, /)
99004 Format ('  Model Parameters: ', /, '     Epsilon: ', F5.3, /, 
     &        '     Sigma  : ', F5.3, /, '     Mass   : ', F5.3)
99007 Format (' Requested Density: ', F5.2, 
     &        ' Different From Density On Disk: ', F5.2, /, 
     &        ' Rescaling Of Coordinates!')
      End

File: Exercises2024/Chapter06_vle_free_energy/C/1-Gibbs/Source/readdat.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>
#include <math.h>
#include "system.h"
#include "potential.h"
#include "conf.h"


void Readdat(int *Equil,int *Prod,int *Nsamp,int *Ndispl,double *Dr,int *Nvol,double *Vmax,int *Nswap,double *Succ)
{
  // Reads Input Data And Model Parameters
  //
  // ---Input Parameters: File: Fort.15
  // BoxIDeg  =  0 : Initilaize From A Lattice
  //          1 : Read Configuration From Disk
  // Equil      : Number Of Monte Carlo Cycles During Equilibration
  // Prod       : Number Of Monte Carlo Cycles During Production
  // Nsamp      : Number Of Monte Carlo Cycles Between Two Sampling Periods
  // Dr         : Maximum Displacement
  // Vmax       : Maximum Volume Change
  // Succ       : Optimal Percentance Of Accepted Attemps
  //              The Program Adjusts Vmax Or Dr In Just A Way That
  //              On Average Succ% Of The Moves Are Accepted
  // Ndispl     : Number Of Attemps To Displace A Particle Per Mc Cycle
  // Nvol       : Number Of Attemps To Change The Volume  Per Mc Cycle
  // Nswap      : Number Of Attemps To Swap Particle Between The Two Boxes Per Mc Cycle
  // Npart      : Total Number of Particles
  // Temp       : Temperature
  // Rho        : Density
  
  //  ---Input Parameters: File: Fort.25
  // Eps    = Epsilon Lennard-Jones Potential
  // Sig    = Sigma Lennard-Jones Potential
  // Mass   = Mass Of The Particle
  // Rcc    = Cut-Off Radius Of The Potential
  
  //  ---Input Parameters: File: Fort.11 (Restart File
  //             To Continue A Simulation From Disk)
  // Box(1)   = Length Box 1 Old Configuration
  // Hbox(1)  = Box(1)/2
  // Box(2)   = Length Box 2 Old Configuration
  // Hbox(2)  = Box(2)/2
  // Npart    = Total Number Of Particles (overrules fort.15!!)
  // Npbox(1) = Number Of Particles In Box 1
  // Npbox(2) = Number Of Particles In Box 2
  // Dr     = Optimized Maximum Displacement Old Configurations
  // Vmax   = Optimized Maximum Volume Change Old Configurations
  // X(1),Y(1),Z(1)            : Position First Particle 1
  //               ,Id(1)   = 1 Particle In Box 1
  //               ,Id(1)   = 2 Particle In Box 2
  //    ....
  // X(Npart),Y(Npart),Z(Npart): Position Particle Last Particle

  int BoxIDeg, I,BoxID;
  double Eps, Sig, Rho, Rcc;
  FILE* fileptr;
  char line[500];
  
  fileptr=fopen("fort.15","r");
  fgets(line,300,fileptr);
  fgets(line,300,fileptr);
  sscanf(line,"%d %d %d %d",&BoxIDeg,Equil,Prod,Nsamp);
  fgets(line,300,fileptr);
  fgets(line,300,fileptr);
  sscanf(line,"%lf %lf %lf",Dr,Vmax,Succ);
  fgets(line,300,fileptr);
  fgets(line,300,fileptr);
  sscanf(line,"%d %d %d",Ndispl,Nvol,Nswap);
  fgets(line,300,fileptr);
  fgets(line,300,fileptr);
  sscanf(line,"%d %lf %lf",&Npart,&Temp,&Rho);
  fclose(fileptr);
  if(Npart>Npmax)
    {
      printf("Error: Number Of Particles Too Large\n");
      exit(0);
    }
  
  //  ---Read Model Parameters
  fileptr=fopen("fort.25","r");
  fgets(line,300,fileptr);
  fgets(line,300,fileptr);
  sscanf(line,"%lf %lf %lf %lf",&Eps,&Sig,&Mass,&Rcc);      
  fclose(fileptr);

  //  ---Read/Generate Configuration
  
  Box[0]  = pow(((double)(Npart)/(2.*Rho)),(1.0/3.0));
  Hbox[0] = 0.5*Box[0];
  Box[1]  = Box[0];
  Hbox[1] = Hbox[0];
  
  if(BoxIDeg==0)
    {
      
      //     ---Generate Configuration Form Lattice
      Lattice();
    }
  else
    {
      printf("Read Conf From Disk\n");
      fileptr=fopen("fort.21","r");
      fscanf(fileptr,"%lf %lf %lf %lf",&(Box[0]), &(Hbox[0]), &(Box[1]), &(Hbox[1]));
      fscanf(fileptr,"%d %d %d",&Npart, &(Npbox[0]), &(Npbox[1]));
      fscanf(fileptr,"%lf %lf",Dr,Vmax);
      for(I = 0;I< Npart;I++)
	{
	  fscanf(fileptr,"%lf %lf %lf %d",&(X[I]),&(Y[I]),&(Z[I]),&(Id[I]));
	}
      //         Rewind (11)
    }
  
  //  ---Write Input Data
  
  printf("Number Of Equilibration Cycles             : %d,\nNumber Of Production Cycles                : %d\nSample Frequency                           :%d\n", *Equil,*Prod,*Nsamp);
  printf("Number Of Att. To Displ. a Part. per Cycle : %d\nMaximum Displacement                       : %lf\nNumber Of Att. to Change Volume  per Cycle : %d\nMaximum Change Volume                      : %lf\nNumber Of Att. to Exch Part.per Cycle    : %d\n",*Ndispl,*Dr,*Nvol,*Vmax,*Nswap);
  printf("Number Of Particles: %d\n",Npart);  
  printf("Temperature: %lf\n",Temp);  
  printf("Pressure: %lf\n",0.0);  
  printf("Density Box 1: %lf\n",(double)(Npbox[0])/pow(Box[0],3));  
  printf("Box 1 Length: %lf\n",Box[0]);  
  printf("Density Box 2: %lf\n",(double)(Npbox[1])/pow(Box[1],3));  
  printf("Box 2 Length: %lf\n",Box[1]);  
  printf("Model Parameters\n");
  printf("Epsilon: %lf\n",Eps);  
  printf("Sigma: %lf\n",Sig);
  printf("Mass: %lf\n",Mass);
  
  //  ---Calculate Parameters:
  
  Beta  = 1.0/Temp;
  Eps4  = 4.0*Eps;
  Eps48 = 48.*Eps;
  Sig2  = Sig*Sig;
  
  //  ---Calculate Cut-Off Radius Potential
  
  for(BoxID = 0;BoxID< 2;BoxID++)
    {
      if(Rcc<Hbox[BoxID])
	{
	  Rc[BoxID] = Rcc;
	}
      else
	{
	  Rc[BoxID]= Hbox[BoxID];
	}
      Rc2[BoxID] = Rc[BoxID]*Rc[BoxID];
    }
  
  return;
}
 
